### PR TITLE
User and frontpage test

### PR DESCRIPTION
**Pull request by EdithHarrison**
**First part of basic testing:**
- The `userController.tests.mjs` file tests the `userController` API, specifically ensuring that the `signup`, `login`, and `logoff` functionalities are working as expected. 
- The `frontpage.tests.mjs` file verifies that the homepage at `http://localhost:5173/` loads correctly.
- To run the tests, navigate to the root directory and type `npm install and npm test` in your terminal.
-**NOTE:** For frontend testing such as frontpage.tests.mjs, you need to launch the app before running the tests.
Screenshot 2024-08-24 at 10 15 41 PM

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -40,6 +49,7 @@
 		"dev:server": "nodemon server/server.mjs",
 		"dev:client": "vite",
 		"build:client": "vite build",
-		"preview:client": "vite preview"
+		"preview:client": "vite preview",
+		"test": "node --experimental-vm-modules node_modules/jest/bin/jest.js"
```
Comment (ElijahMSmith) [inline package.json:53]: Curious about this script. What issues did you run into that require this flag (genuinely curious, not trying to make anyone feel stupid).

The "Getting Started" guide suggests you can use `"test": "jest"`. Does that work here?
https://jestjs.io/docs/getting-started

File: __tests__/userController.tests.mjs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+                token: expect.any(String),
+                user: expect.objectContaining({ uid: expect.any(String) }),
+            }));
+        });
+    });
+
+    // Test user logoff functionality
+    describe('logoffUser', () => {
+        it('should log off a user successfully', async () => {
+            const reqLogoff = mockRequest({
```
Comment (ElijahMSmith) [inline __tests__/userController.tests.mjs:80]: Can we use your helper `createMockRequestResponse` here as well?

File: __tests__/userController.tests.mjs [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+import admin from 'server/config/firebase.mjs';
+
+describe('User Controller', () => {
+    // Array to store all user IDs created during testing
+    const createdUserIds = [];
+    let testEmail;
+    const TEST_PASSWORD = 'password123';
+
+    // Helper function to create mock request and response objects
+    const createMockRequestResponse = (body) => {
```
Comment (ElijahMSmith) [inline __tests__/userController.tests.mjs:12]: Total nitpick so feel free to ignore.

I think extracting this to a helper function makes the tests a little harder to read. It's not really saving us much if any typing, and it's not clear when I call the function from inside a test what the object I'm passing in is for. I would personally choose to explicitly call `mockRequst` and `mockResponse` in each test that wants them.

**Review by ElijahMSmith — DISMISSED** [review]
Really great tests!

File: package.json (hunk 1 above)
Comment (mattschmitz) [reply package.json:53]: also was wondering

File: .babelrc
Comment (mattschmitz) [inline .babelrc:1]: curious about this  - why is it needed? I thought vite meant we didn't need bable. Genuine question.

File: package.json (hunk 1 above)
Comment (EdithHarrison) [reply package.json:53]: Hi Eli and Matthew I use --experimental-vm-modules here because our project relies on ES modules (.mjs files), and without this flag, Jest throws errors like "SyntaxError: Cannot use import statement outside a module". The standard "test": "jest" script doesn’t support ES modules out of the box.

File: __tests__/userController.tests.mjs (hunk 3 above)
Comment (EdithHarrison) [reply __tests__/userController.tests.mjs:12]: I did try to explicitly call mockRequest and mockResponse, but I kept getting a “require is not defined” error. Even though we don’t have require in our code, the mockRequest and mockResponse package is trying to run our code in a CommonJS environment, which expects require to be available. Since we’re using ES modules and import instead, so I used helper function as a workaround.

File: .babelrc
Comment (EdithHarrison) [reply .babelrc:1]: I use Babel here specifically for our tests because I don’t think Vite handles the compatibility issues needed for testing with Jest. I’m not sure though I could be wrong.

File: __tests__/frontpage.tests.mjs
Comment (SonaliBedge) [inline __tests__/frontpage.tests.mjs:1]: Just a suggestion: consider using a try/catch block. You can choose to ignore it if you prefer.

File: __tests__/userController.tests.mjs (hunk 2 above)
Comment (SonaliBedge) [reply __tests__/userController.tests.mjs:80]: if possible try to add more test cases like invalid input/user and duplicate user.

**Review by SonaliBedge — DISMISSED** [review]

**Review by OGoncharenko — DISMISSED** [review]

**Review by mattschmitz — DISMISSED** [review]
spoke in person. Please have a peer test and then tag me.

**Review by SonaliBedge — APPROVED** [review]
https://codethedream.slack.com/files/U05Q9EC5DAT/F07KE67GXDG/image.png

File: __tests__/userController.tests.mjs [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    }, 30000);
+
+    describe('signupUser', () => {
+        it('should sign up a new user successfully', async () => {
+            expect(createdUserIds[createdUserIds.length - 1]).toBeDefined();
+        });
+    });
+
+    // Test user login functionality
+    describe('loginUser', () => {
```
[the remark this answers is not in the document]
Comment (EdithHarrison) [reply __tests__/userController.tests.mjs:59]: Hi Matthew. Sonali run the tests for me and all passed.

File: .babelrc [hunk 5]
```diff
@@ -0,0 +1,3 @@
+{
```
[the remark this answers is not in the document]
Comment (EdithHarrison) [reply .babelrc]: Removed

**Review by SonaliBedge — DISMISSED** [review]

**Review by SonaliBedge — DISMISSED** [review]

**Review by mattschmitz — APPROVED** [review]
Didn't have time for proper re-review, but quick scroll and I don't see any major issues. Typically you would not have separate npm `test` scripts for every module - you would just have maybe one overall, one for frontend and one for backend. But not a problem, just letting you know.
